### PR TITLE
Store synchronization with shared reducer

### LINKED ISSUE #4: Implement redux store synchronization on the frontend

### LINKED ISSUE #5: Implement redux store synchronization on the backend

### PR DESCRIPTION
**Pull request by IlyaChebanu**
closes #5 
closes #4

### CODE REVIEW

File: api_src/server.js [hunk 1]
```diff
@@ -12,11 +12,16 @@ const store = configureStore(sagaMiddleware);
...
-store.dispatch(userUploadActions.test("dingus amingus"))
+setTimeout(() => {
+    for (let i = 0; i < 10; i++) {
+        store.dispatch(userUploadActions.test("dingus amingus"))
+        store.dispatch(userUploadActions.test("dingus amingus"))
+    }
+}, 5000);
 console.log("testing my logs lmao")
 
 app.get("/", (req, res) => {
```
Comment (sss1612) [inline api_src/server.js:23]: So how is the webpage going to be served if not via base url

File: api_src/store/synchronizer.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
+import WebSocket from 'ws';
+
+const wss = new WebSocket.Server({ port: 7070 });
+
+wss.on('connection', ws => {
+  ws.messageId = 0;
```
Comment (sss1612) [inline api_src/store/synchronizer.js:6]: Just currently allowing 1 client initially? Think there needs to be something in the store later on to keep track of the id's

File: api_src/store/synchronizer.js [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+});
+
+let subscribedStores = [];
```
Comment (sss1612) [inline api_src/store/synchronizer.js:9]: STOOOOOOOOOOOOOOOOOOOORE. PUT IT IN A STOOOOOOOOOOOOORE

File: api_src/store/synchronizer.js [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+const synchronizer = store => next => action => {
```
Comment (sss1612) [inline api_src/store/synchronizer.js:10]: :(

Comment (sss1612) [reply api_src/store/synchronizer.js:10]: It's not obvious what's happening here. Is every action being directed here?

File: src/App.css [hunk 5]
```diff
@@ -9,7 +9,7 @@
 
 @media (prefers-reduced-motion: no-preference) {
   .App-logo {
-    animation: App-logo-spin infinite 0.01s linear;
+    animation: App-logo-spin infinite 10s linear;
```
Comment (sss1612) [inline src/App.css:12]: 😢

File: shared/store/userUpload.js [hunk 6]
```diff
@@ -0,0 +1,55 @@
+
```
Comment (sss1612) [inline shared/store/userUpload.js:1]: Gone through this with you 👍

File: src/App.js [hunk 7]
```diff
@@ -2,7 +2,7 @@ import React from 'react';
 import logo from './logo.svg';
 import './App.css';
 import { connect } from "react-redux";
-import { selectors as userUploadSelectors } from "./store/userUpload/userUpload";
+import { selectors as userUploadSelectors } from "../shared/store/userUpload";
```
Comment (sss1612) [inline src/App.js:5]: 🔨

File: src/store/rootReducer.js [hunk 8]
```diff
@@ -1,7 +1,8 @@
 import { combineReducers  } from "redux";
-import userUpload from "./userUpload/userUpload";
+import userUpload from "../../shared/store/userUpload";
 
 const rootReducer = combineReducers({
     userUpload,
-})
+});
+
```
Comment (sss1612) [inline src/store/rootReducer.js:7]: Ty

File: src/store/synchronizer.js [hunk 9]
```diff
@@ -0,0 +1,30 @@
+let subscribedStores = [];
```
Comment (sss1612) [inline src/store/synchronizer.js:1]: Not tryna be captain obvious but basically the client store here is diff to the server store and every action is logged? I'm still unsure about line 9 in this file

File: api_src/store/synchronizer.js (hunk 3 above)
Comment (sss1612) [reply api_src/store/synchronizer.js:9]: Wait, probably doesn't need to be yet. Idk

File: api_src/server.js (hunk 1 above)
Comment (IlyaChebanu) [reply api_src/server.js:23]: dunno, haven't done ssr

File: api_src/store/synchronizer.js (hunk 2 above)
Comment (IlyaChebanu) [reply api_src/store/synchronizer.js:6]: no, this runs for every client and adds it to wss.clients, so every item in clients will have a ws.messageId thats diff to every other

File: api_src/server.js (hunk 1 above)
Comment (sss1612) [reply api_src/server.js:23]: Don't think it's a matter of ssr. We just need to build before we serve the webpage

File: api_src/store/synchronizer.js (hunk 3 above)
Comment (IlyaChebanu) [reply api_src/store/synchronizer.js:9]: this has no business living in the store, this is just to locally track what we want to sync

File: api_src/server.js (hunk 1 above)
Comment (sss1612) [reply api_src/server.js:23]: If that is ssr, excuse the retardary

File: api_src/store/synchronizer.js (hunk 4 above)
Comment (IlyaChebanu) [reply api_src/store/synchronizer.js:10]: this is just the format of redux middleware, and yeah every single action comes into it

File: src/store/synchronizer.js (hunk 9 above)
Comment (IlyaChebanu) [reply src/store/synchronizer.js:1]: no idea what you're talking about m80

File: api_src/store/synchronizer.js (hunk 2 above)
Comment (sss1612) [reply api_src/store/synchronizer.js:6]: Ah. My eyes are going. `wss` and `ws`...

**Pull request merged**